src/lua/protobuf.lua

```diff
@@ -0,0 +1,316 @@
+        end
+    end
+    setmetatable(return_protocol, protocol_mt)
+    return return_protocol
+end
+
+local function check_integer(data_spec, value)
+    if type(value) == 'number' and math.ceil(value) ~= value then
+        error(('Input number value %f for %q is not integer'):format(
+            value, data_spec['name']))
+    elseif (type(value) == 'cdata' and
+            not c_integer[tostring(ffi.typeof(value))]) then
+        error(('Input cdata value %q for %q field is not integer'):format(
+            tostring(ffi.typeof(value)), data_spec['name']))
+    end
+end
+
+local function code(data_spec, value)
+    local data_type = data_spec['type']
+    if (protobuf_varint_types[data_type] and
+        input_varint_types[type(value)]) then
+        check_integer(data_spec, value)
+        return encoder.encode_integer(data_spec, value)
+    elseif (protobuf_len_types[data_type] and
+            input_len_types[type(value)]) then
+        return encoder.encode_len(data_spec, value)
+    elseif (data_type == 'float' and type(value) == 'number') then
+        return encoder.encode_I32(data_spec, value)
+    elseif (data_type == 'double' and type(value) == 'number') then
+        return encoder.encode_I64(data_spec, value)
+    elseif (protobuf_I32_types[data_type] and
+            input_I32_types[type(value)]) then
+        if (type(value) == 'number' and (value > max_int32-1 or
+                value < -max_int32)) then
+            error(('Input number value %q for %q field does ' ..
+                'not fit in int32'):format(value, data_spec['name']))
+        end
+        check_integer(data_spec, value)
+        return encoder.encode_I32(data_spec, value)
+    elseif (protobuf_I64_types[data_type] and
+            input_I64_types[type(value)]) then
+        if (type(value) == 'number' and (value >= max_int64 or
+                value < -max_int64)) then
+            error(('Input number value %q for %q field does ' ..
+                'not fit in int64'):format(value, data_spec['name']))
+        end
+        check_integer(data_spec, value)
+        return encoder.encode_I64(data_spec, value)
+    else
+        error(('Field %q of %q type gets %q type value. ' ..
+            'Unsupported or colliding types'):format(data_spec['name'],
+            data_type, type(value)))
+    end
+end
+
+-- Encodes the entered data in accordance with the
+-- selected protocol into binary format.
+--
+-- Accepts a protocol created by protobuf.protocol function,
+-- a name of a message selected for encoding and
+-- the data that needs to be encoded in the following format.
+--
+-- data = {
+--     <field_name> = <value>,
+--     <...>
+-- }
+local function encode(protocol, name, data)
+    local coded_data = {}
+    local unknown_fields = {}
+    local unknown_fields_counter = 0
+    local message_protocol = protocol[name]
+    if message_protocol == nil then
+        error(('Wrong message name %q for this protocol'):format(name))
+    end
```

Comment:
Nit: I would say that there is no such message or enum name in the given protocol. The error looks like the name is ill-formed (has forbidden symbols) or something of this kind.

Reply:
Changed error message to "There is no message or enum named %q in the given protocol"